### PR TITLE
[WebDriver BiDi] Add screenshot origin/format tests

### PR DESCRIPTION
**Pull request by jrandolf**
Tests for https://github.com/w3c/webdriver-bidi/pull/564 and https://github.com/w3c/webdriver-bidi/pull/561

### CODE REVIEW

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 1]
```diff
@@ -35,9 +35,16 @@ def activate(self, context: str) -> Mapping[str, Any]:
...
     def capture_screenshot(
-        self, context: str, clip: Optional[ClipOptions] = None
-    ) -> Mapping[str, Any]:
-        params: MutableMapping[str, Any] = {"context": context}
+            self,
+            context: str,
+            format: Optional[str] = None,
+            origin: Optional[str] = None,
+            clip: Optional[ClipOptions] = None) -> Mapping[str, Any]:
+        params: MutableMapping[str, Any] = {
```
Comment (OrKoN) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:43]: I think we should only set keys if the value is not None. None would serialize as null which is not what we want.

**Review by lutien — COMMENTED** [review]
I see that the target for this PR is [another PR](https://github.com/web-platform-tests/wpt/tree/jrandolf/viewport_unchanged) and not master, could you maybe change it? It's problematic because the changes from the target PR cause the unit test failures here.

Also, I guess we should test the combination of `origin=document` and `clip`, especially interesting would be the case when the target area is outside of viewport. What do you think?

File: webdriver/tests/bidi/browsing_context/capture_screenshot/format.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        wait="complete")
+
+    png_screenshot = await bidi_session.browsing_context.capture_screenshot(
+        context=top_context["context"], format={"type": "image/png"})
+    jpeg_screenshot = await bidi_session.browsing_context.capture_screenshot(
+        context=top_context["context"], format={"type": "image/jpeg"})
+    jpeg_quality_screenshot = await bidi_session.browsing_context.capture_screenshot(
+        context=top_context["context"],
+        format={
+            "type": "image/jpeg",
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/format.py:18]: There is only one other image type checked here. Which other types should be supported by default?

Also can we somewhat test different quality settings? I assume we could verify by just comparing the size of the base64 encoded data?

Comment (jrandolf) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/format.py:18]: The only image type required to be supported in the HTML standard is `image/png` (https://html.spec.whatwg.org/multipage/canvas.html#a-serialisation-of-the-bitmap-as-a-file)

I've added a test for quality.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 3]
```diff
@@ -41,9 +41,15 @@ def activate(self, context: str) -> Mapping[str, Any]:
     def capture_screenshot(
             self,
             context: str,
+            format: Optional[Mapping[str, Any]] = None,
+            origin: Optional[str] = None,
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:45]: Could you please add custom types similar to `ClipOptions`? Thanks.

**Review by whimboo — CHANGES_REQUESTED** [review]
Can you please fix the linter failures in the various jobs?

Note that this PR uses the `UNDEFINED` type which hasn't been landed yet and awaits a merge on #42913. As such we cannot merge it yet.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import pytest
+
+from tests.support.image import png_dimensions
+
+from . import get_physical_viewport_dimensions, get_physical_document_dimensions
+
+
+@pytest.mark.asyncio
+async def test_origin(bidi_session, top_context, inline):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py:9]: @jrandolf it would be great if you could parametrize this test so that both settings can be tested individually and an early failure of `document` as origin will not prevent a check for `viewport`.  Ma

File: webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py [hunk 5]
```diff
@@ -124,3 +124,22 @@ async def test_set_viewport_persists_on_reload(bidi_session, new_tab):
     )
 
     assert await get_viewport_dimensions(bidi_session, new_tab) == test_viewport
+
+
+@pytest.mark.asyncio
+async def test_set_viewport_undefined(bidi_session, new_tab):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/set_viewport/set_viewport.py:130]: These changes are not related to screenshots. Please separate them out to the right PR which is most likely #42713.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 from ... import get_viewport_dimensions
 
 import webdriver.bidi.error as error
-from webdriver.bidi.modules.browsing_context import ElementOptions, BoxOptions
+from webdriver.bidi.modules.browsing_context import ElementOptions, BoxOptions, FormatOptions
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py:6]: While you are at it:

```suggestion
from webdriver.bidi.modules.browsing_context import BoxOptions, ElementOptions, FormatOptions
```

File: webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py [hunk 7]
```diff
@@ -21,41 +21,40 @@ async def test_invalid_frame(bidi_session, value):
         await bidi_session.browsing_context.capture_screenshot(context=value)
 
 
-async def test_closed_frame(bidi_session, top_context, inline, add_and_remove_iframe):
+async def test_closed_frame(bidi_session, top_context, inline,
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py:24]: Mind reverting the formatting changes to make it easier to see the actual important changes? Thanks.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py [hunk 8]
```diff
@@ -123,28 +122,28 @@ async def test_params_clip_viewport_width_invalid_type(
...
             context=top_context["context"],
             clip=BoxOptions(x=0, y=0, width=0, height=0),
         )
 
 
-async def test_params_clip_viewport_outside_of_window_viewport(
-    bidi_session, top_context
-):
-    viewport_dimensions = await get_viewport_dimensions(bidi_session, top_context)
+async def test_params_clip_box_outside_of_window_viewport(
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py:143]: I would suggest to move the tests that do not test for `invalid argument` to the corresponding test module. In this case it looks like to be related to `clip.py`. This also applies to the next one.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py [hunk 9]
```diff
@@ -182,3 +180,16 @@ async def test_params_clip_element_outside_of_window_viewport(
             context=top_context["context"],
             clip=ElementOptions(element=element),
         )
+
+
+async def test_params_origin_invalid(bidi_session, top_context):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/invalid.py:185]: Please add a test before for invalid types. The same applies for `format` as well.

```suggestion
async def test_params_origin_invalid_value(bidi_session, top_context):
```

File: webdriver/tests/bidi/browsing_context/capture_screenshot/format.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        url=inline("<div style='margin-top:2000px'>foo</div>"),
+        wait="complete")
+
+    png_screenshot = await bidi_session.browsing_context.capture_screenshot(
+        context=top_context["context"]
+        format=FormatOptions(type="image/png"))
+    jpeg_screenshot = await bidi_session.browsing_context.capture_screenshot(
+        context=top_context["context"],
+        format=FormatOptions(type="image/jpeg"))
+    jpeg_quality_screenshot = await bidi_session.browsing_context.capture_screenshot(
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/format.py:18]: Please move the quality tests to a separate test.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/format.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import pytest
+
+from webdriver.bidi.modules.browsing_context import FormatOptions
+
+@pytest.mark.asyncio
+async def test_format(bidi_session, top_context, inline):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/format.py:6]:
```suggestion
async def test_format_type(bidi_session, top_context, inline):
```

File: tools/webdriver/webdriver/undefined.py [hunk 12]
```diff
@@ -0,0 +1,6 @@
+class Undefined:
```
Comment (whimboo) [inline tools/webdriver/webdriver/undefined.py:1]: These changes are for PR #42864. Lets make sure we wait for its merge first.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 13]
```diff
@@ -180,16 +203,16 @@ def _print(self, result: Mapping[str, Any]) -> Any:
     @command
     def set_viewport(self,
                      context: str,
-                     viewport: Optional[Mapping[str, Any]] = None,
-                     device_pixel_ratio: Optional[float] = None) -> Mapping[str, Any]:
+                     viewport: Union[Optional[Mapping[str, Any]], Undefined] = UNDEFINED,
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:206]: Those changes are as well not for this PR. Please remove.

**Review by whimboo — CHANGES_REQUESTED** [review]

**Comment by whimboo** [comment]
@jrandolf we added the feature to Firefox yesterday. To have some internal tests as long as your PR hasn't been merged we picked the parts you added to the WebDriver BiDi client and already landed those via https://github.com/web-platform-tests/wpt/pull/43003/files.

Please rebase your branch when you continue. Thanks.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py (hunk 4 above)
Comment (jrandolf) [reply webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py:9]: We test the document screenshot by comparing it with the viewport screenshot.

**Comment by whimboo** [comment]
@jrandolf mind resolving the conflict for this PR before I do the next review? Tests for `origin` look fine so far and pass all in Firefox. Thanks.

**Comment by jrandolf** [comment]
> @jrandolf mind resolving the conflict for this PR before I do the next review? Tests for `origin` look fine so far and pass all in Firefox. Thanks.

@whimboo Done.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 14]
```diff
@@ -30,6 +30,13 @@ class OriginOptions(Enum):
     DOCUMENT = "document"
     VIEWPORT = "viewport"
 
+class FormatOptions(Dict[str, Any]):
+
+    def __init__(self, type: str, quality: Optional[float] = None):
+        self["type"] = type
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:36]: We should correctly initialize the class as dict, similar to the other custom classes in other modules:

```suggestion
    def __init__(self, type: str, quality: Optional[float] = None):
        dict.__init__(self, type=type)

```

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 15]
```diff
@@ -30,6 +30,13 @@ class OriginOptions(Enum):
     DOCUMENT = "document"
     VIEWPORT = "viewport"
 
+class FormatOptions(Dict[str, Any]):
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:33]:
```suggestion


class FormatOptions(Dict[str, Any]):
```

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 16]
```diff
@@ -38,16 +45,19 @@ def activate(self, context: str) -> Mapping[str, Any]:
 
     @command
     def capture_screenshot(
-        self,
-        context: str,
-        clip: Optional[ClipOptions] = None,
-        origin: Optional[OriginOptions] = None,
-    ) -> Mapping[str, Any]:
+            self,
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:48]: I know that this is indentation again but we use 4 chars or align it by the opening bracket the line before. This (8 chars) is unusual, and does flake8 complain?

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 17]
```diff
@@ -38,16 +45,19 @@ def activate(self, context: str) -> Mapping[str, Any]:
 
     @command
     def capture_screenshot(
-        self,
-        context: str,
-        clip: Optional[ClipOptions] = None,
-        origin: Optional[OriginOptions] = None,
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:44]: Why did you revert `OriginOptions`? We have a dedicated type for that.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 18]
```diff
@@ -38,16 +45,19 @@ def activate(self, context: str) -> Mapping[str, Any]:
...
+            context: str,
+            format: Optional[FormatOptions] = None,
+            origin: Optional[str] = None,
+            clip: Optional[ClipOptions] = None) -> Mapping[str, Any]:
         params: MutableMapping[str, Any] = {"context": context}
 
+        if format is not None:
+            params["format"] = format
+        if origin is not None:
+            params["origin"] = origin
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:58]: It's already present below so please remove these two lines.
```suggestion
```

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 19]
```diff
@@ -38,16 +45,19 @@ def activate(self, context: str) -> Mapping[str, Any]:
...
         if clip is not None:
             params["clip"] = clip
-
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:50]: nit: We always have an empty line in other methods.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/clip.py [hunk 20]
```diff
@@ -1,4 +1,6 @@
 import pytest
+
+import webdriver.bidi.error as error
 from tests.support.image import png_dimensions
 from webdriver.bidi.modules.browsing_context import ElementOptions, BoxOptions
 from webdriver.bidi.modules.script import ContextTarget
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/clip.py:6]: `webdriver` is still a third-party package while tests can be considered local imports. Maybe you can fix that as ride-along?

```suggestion
import webdriver.bidi.error as error
from webdriver.bidi.modules.browsing_context import ElementOptions, BoxOptions
from webdriver.bidi.modules.script import ContextTarget

from tests.support.image import png_dimensions

```

File: webdriver/tests/bidi/browsing_context/capture_screenshot/clip.py [hunk 21]
```diff
@@ -260,3 +258,46 @@ async def test_clip_viewport_partially_visible(
...
+        await_promise=False,
+        expression="document.querySelector('div')",
+        target=ContextTarget(top_context["context"]),
+    )
+
+    with pytest.raises(error.UnableToCaptureScreenException):
+        await bidi_session.browsing_context.capture_screenshot(
+            context=top_context["context"],
+            clip=ElementOptions(element=element),
+        )
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/clip.py:303]: nit: Please add an empty line at the end. Same for other files as well.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/format.py [hunk 22]
```diff
@@ -0,0 +1,38 @@
+import pytest
+
+from webdriver.bidi.modules.browsing_context import FormatOptions
+
+@pytest.mark.asyncio
+async def test_format_type(bidi_session, top_context, inline):
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/format.py:6]:
```suggestion


@pytest.mark.asyncio
async def test_format_type(bidi_session, top_context, inline):
```

**Review by whimboo — CHANGES_REQUESTED** [review]
Some remaining comments. Otherwise it seems to be fine now.

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py [hunk 23]
```diff
@@ -7,9 +7,7 @@
 
 
 class ElementOptions(Dict[str, Any]):
-    def __init__(
-        self, element: Mapping[str, Any]
-    ):
+    def __init__(self, element: Mapping[str, Any]):
```
Comment (whimboo) [inline tools/webdriver/webdriver/bidi/modules/browsing_context.py:10]: Looks like we missed the call to `dict.__init__()` here as well, but up to you to fix it now or later.

File: webdriver/tests/bidi/browsing_context/capture_screenshot/__init__.py [hunk 24]
```diff
@@ -4,6 +4,7 @@
     get_element_dimensions,
     get_viewport_dimensions,
     remote_mapping_to_dict,
+    get_document_dimensions,
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/__init__.py:7]:
```suggestion
    get_document_dimensions,
    get_element_dimensions,    
    get_viewport_dimensions,
    remote_mapping_to_dict,
```

File: webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
+import pytest
+
+from tests.support.image import png_dimensions
+
+from . import get_physical_viewport_dimensions, get_physical_document_dimensions
+
```
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/capture_screenshot/origin.py:6]:
```suggestion
from . import get_physical_document_dimensions, get_physical_viewport_dimensions

```

**Review by whimboo — APPROVED** [review]
Thank you for the update @jrandolf! It looks all fine, just some nits.

**Review by OrKoN — APPROVED** [review]

File: tools/webdriver/webdriver/bidi/modules/browsing_context.py (hunk 23 above)
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/browsing_context.py:10]: @jrandolf please follow up on this once you are back
